File: BPG-Bioland/BPGBLE/BPGProtocol.m
//
//  BPGProtocol.m
//  BPG-Bioland
//
//  Created by starlueng on 2016/8/10.
//  Copyright © 2016年 starlueng. All rights reserved.
//


#import "BPGProtocol.h"
#import "MBProgressHUD.h"
#import "NSDate+SSToolkitAdditions.h"
static NSString *const channelOnCharacteristicView = @"babyDefault";
static NSString *const serverID                   = @"0000FBB0-494C-4F47-4943-544543480000";//服务器id
static NSString *const readCharacterID             = @"0000FBB1-494C-4F47-4943-544543480000";//读特征值
static NSString *const writeCharacterID            = @"0000FBB2-494C-4F47-4943-544543480000";//写特征值
@implementation BPGDataModel

@end
@implementation BPGProtocol

+ (instancetype) shareInstance{
    
    static BPGProtocol  *protocol = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        protocol = [[BPGProtocol alloc]init];
    });
    return protocol;
}
- (void)scanBleWithPrexName:(NSString *)prexName{
    //停止之前的连接
    [self.baby cancelAllPeripheralsConnection];
    //设置委托后直接可以使用，无需等待CBCentralManagerStatePoweredOn状态。
    self.baby.scanForPeripherals().begin();
    
    __weak typeof(self) weakSelf = self;
    
    [self.baby setBlockOnCentralManagerDidUpdateState:^(CBCentralManager *central) {
        if (central.state == CBCentralManagerStatePoweredOn) {
            NSLog(@"蓝牙已打开");
        }
    }];
    //1.设置查找设备的过滤器
    [self.baby setFilterOnDiscoverPeripherals:^BOOL(NSString *peripheralName, NSDictionary *advertisementData, NSNumber *RSSI) {
        //最常用的场景是查找某一个前缀开头的设备
        if ([[peripheralName lowercaseString] hasPrefix:prexName] ) {
            
            NSLog(@"找到设备:%@",peripheralName);
            return YES;
        }
        return NO;
    }];
    
    //2.设置扫描到设备的委托
    [self.baby setBlockOnDiscoverToPeripherals:^(CBCentralManager *central, CBPeripheral *peripheral, NSDictionary *advertisementData, NSNumber *RSSI) {
        NSLog(@"搜索到了设备:%@",peripheral.name);
        if ([[peripheral.name lowercaseString] hasPrefix:prexName] ) {
            
            [weakSelf.baby cancelScan];
            weakSelf.currPeripheral = peripheral;
            MBProgressHUD *hud =[MBProgressHUD showHUDAddedTo:[[[UIApplication sharedApplication]delegate]window ] animated:YES];
            [weakSelf connectBleWith:^(CBCharacteristic *writer, CBCharacteristic *read) {
                
                [hud hide:YES];
            }];
        }
    }];
    //3.扫描选项->CBCentralManagerScanOptionAllowDuplicatesKey:忽略同一个Peripheral端的多个发现事件被聚合成一个发现事件
    NSDictionary *scanForPeripheralsWithOptions = @{CBCentralManagerScanOptionAllowDuplicatesKey:@YES};
    /*连接选项->
     CBConnectPeripheralOptionNotifyOnDisconnectionKey :当应用挂起时，如果连接断开时，如果我们想要系统为指定的peripheral显示一个断开连接的提示时，就使用这个key值。
     CBConnectPeripheralOptionNotifyOnNotificationKey:
     当应用挂起时，使用该key值表示只要接收到给定peripheral端的通知就显示一个提
     */
    NSDictionary *connectOptions = @{CBConnectPeripheralOptionNotifyOnConnectionKey:@YES,
                                     CBConnectPeripheralOptionNotifyOnDisconnectionKey:@YES,
                                     CBConnectPeripheralOptionNotifyOnNotificationKey:@YES
                                     };
    //连接设备->
    [self.baby setBabyOptionsAtChannel:channelOnCharacteristicView scanForPeripheralsWithOptions:scanForPeripheralsWithOptions connectPeripheralWithOptions:connectOptions scanForPeripheralsWithServices:nil discoverWithServices:nil discoverWithCharacteristics:nil];

}
- (void)loadData{
    
    //4.开始连接设备
    self.baby.having(self.currPeripheral).and.channel(channelOnCharacteristicView).then.connectToPeripherals().discoverServices().discoverCharacteristics().readValueForCharacteristic().discoverDescriptorsForCharacteristic().readValueForDescriptors().begin();
}
- (void)connectBleWith:(Block)success{
    //开始扫描设备
    [self performSelector:@selector(loadData) withObject:nil afterDelay:0.5];
    

    BabyRhythm *rhythm = [[BabyRhythm alloc]init];
    
    __weak typeof(self) weakSelf = self;
    
    //设置发现设service的Characteristics的委托
    [self.baby setBlockOnDiscoverCharacteristicsAtChannel:channelOnCharacteristicView block:^(CBPeripheral *peripheral, CBService *service, NSError *error) {
        NSLog(@"===service name:%@",service.UUID);
        [rhythm beats];
    }];
    [self.baby setBlockOnDisconnect:^(CBCentralManager *central, CBPeripheral *peripheral, NSError *error) {
        NSLog(@"didDisconnectPeripheral-----%@",error);
        [weakSelf loadData];
    }];
    
    //设置读取characteristics的委托
    [self.baby setBlockOnReadValueForCharacteristicAtChannel:channelOnCharacteristicView block:^(CBPeripheral *peripheral, CBCharacteristic *characteristics, NSError *error) {
        //获取设备的mac地址
        if ([characteristics.UUID.UUIDString isEqualToString:@"2A23"]) {
            
            weakSelf.information = characteristics.value;
        }
        if ([characteristics.UUID.UUIDString isEqualToString:writeCharacterID]) {
            
            weakSelf.writerCharacteristic = characteristics;
            
        }else if ([characteristics.UUID.UUIDString isEqualToString:readCharacterID]){
            weakSelf.readCharacteristic = characteristics;
            

        }
        if (weakSelf.writerCharacteristic&&weakSelf.readCharacteristic) {
            success(weakSelf.writerCharacteristic,weakSelf.readCharacteristic);
        }
        
    }];
    
    //设置写数据成功的block
    [self.baby setBlockOnDidWriteValueForCharacteristicAtChannel:channelOnCharacteristicView block:^(CBCharacteristic *characteristic, NSError *error) {
        NSLog(@"setBlockOnDidWriteValueForCharacteristicAtChannel characteristic:%@ and new value:%@",characteristic.UUID, characteristic.value);
    }];
   
}
//蓝牙网关初始化和委托方法设置
-(void)babyDelegate{
    //搜索蓝牙
    [self scanBleWithPrexName:@"serialcom"];
}

- (BabyBluetooth *)baby{
    
    if (!_baby) {
        
        _baby = [BabyBluetooth shareBabyBluetooth];
    }
    return _baby;
}

- (BPGDataModel *)transFormData:(NSData *)data{
    
    uint8_t packetData[data.length] ;
    memcpy(packetData, data.bytes, data.length);
    
   
    if (packetData[0] !=0x68) {
        
        return nil;
    }else{
        BPGDataModel *model = [[BPGDataModel alloc]init];
        model.head              = packetData[0];
        model.cmd               = packetData[1];
        model.status            = packetData[2];
        NSInteger bodyLength    = (packetData[4]<<8) +packetData[5];
        model.leng              = bodyLength;
        uint8_t bodys[bodyLength];
        memcpy(bodys, packetData+5, bodyLength);
        model.body              =[NSData dataWithBytes:&bodys length:bodyLength];
        model.checkcs           = packetData[data.length-2];
        model.end               = packetData[data.length-1];
        
        return model;
    }
}

-(void)writeValueWith:(NSData *)data AndNotifiy:(getResult)success OrFail:(getResult)fail{

    if (self.currPeripheral.state ==CBPeripheralStateConnected) {
        //写一个值
        [self.currPeripheral writeValue:data forCharacteristic:self.writerCharacteristic type:CBCharacteristicWriteWithResponse];
        //订阅一个值
        __weak typeof(self)weakSelf = self;
        if (!self.currPeripheral) {
            NSError *error = [NSError errorWithDomain:@"蓝牙设备未连接" code:0 userInfo:nil];
            fail(error);
           
            return;
        }
        if(self.currPeripheral.state != CBPeripheralStateConnected) {
            NSError *error = [NSError errorWithDomain:@"蓝牙已经断开连接，请重新连接" code:0 userInfo:nil];
            fail(error);

            return;
        }
        if (self.readCharacteristic.properties & CBCharacteristicPropertyNotify ||  self.readCharacteristic.properties & CBCharacteristicPropertyIndicate) {
            
            if(!self.readCharacteristic.isNotifying) {
                
                [weakSelf.currPeripheral setNotifyValue:YES forCharacteristic:self.readCharacteristic];
                
                [self.baby notify:self.currPeripheral
                   characteristic:self.readCharacteristic
                            block:^(CBPeripheral *peripheral, CBCharacteristic *characteristics, NSError *error) {
                                NSLog(@"%@",characteristics.value);
                                BPGDataModel *model = [weakSelf transFormData:characteristics.value];
                                
                                success(model);
                            }];
            }
        }
        else{
            NSError *error = [NSError errorWithDomain:@"这个characteristic没有nofity的权限" code:0 userInfo:nil];
            fail(error);
           
            return;
        }
    }
}

- (void)babyStopConnect{

    [self.baby cancelScan];
    //蓝牙断开连接
    [self.baby cancelAllPeripheralsConnection];
}
@end
@implementation BPGProtocol (Commond)
- (NSData *)writeCommodDataWithcmd:(uint8_t)cmd Andcommond:(commond)common AndbodyData:(NSData *)body{
    NSInteger dataLength = 9+body.length;
    NSInteger bodyLength = 2+body.length;
    
    char byte[dataLength];
    memset(byte, 0, dataLength);
    byte[0] = 0x68;
    byte[1] = cmd;
    byte[3] = bodyLength%256;
    byte[4] = bodyLength/256;
    byte[5] = common>>8;
    byte[6] = common&0x00ff;
    if (body.length>0) {
        
        uint8_t bodys[body.length];
        memcpy(bodys, body.bytes, body.length);
        memcpy(byte+7, bodys, body.length);
    }

    NSInteger calibrate =0;
    for (NSInteger i=0; i<dataLength-2; i++) {
        calibrate += byte[i];
    }
    
    byte[dataLength-2] = calibrate&0xff;
    byte[dataLength-1] = 0x16;
    return [NSData dataWithBytes:byte length:dataLength];
}
- (void)getLastDataWith:(getResult )result OrFail:(getResult)fail{

    Byte bodyByte[] ={};
    
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:getLastTestData AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)realTestMode:(getResult)result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:realTestMode AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
         result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)pauseMode:(getResult)result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:pauseMode AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)getAllTestDataWith:(getResult )result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:getAllTestData AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];
}
- (void)turnOnModeWith:(getResult )result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:turnOnMode AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)turnOffModeWith:(getResult )result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:turnOffMode AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)setTimeWithTimeData:(NSString *)time result:(getResult )result OrFail:(getResult)fail{
    
    /*   如果不传time 默认设置时间为当前时间，如果传time则是设置时间   */
    NSData *valueData = [self writeCommodDataWithcmd:0x06 Andcommond:setTime AndbodyData:[self setTimeDataWith:nil]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}
- (void)isSendDeployWith:(getResult )result OrFail:(getResult)fail{
    Byte bodyByte[] ={};
    NSData *valueData = [self writeCommodDataWithcmd:0x02 Andcommond:isSendDeploy AndbodyData:[NSData dataWithBytes:bodyByte length:0]];
    [self writeValueWith:valueData AndNotifiy:^(BPGDataModel *model) {
        
        result(model);
    } OrFail:^(NSError *error) {
        fail(error);
    }];

}

- (NSData *)setTimeDataWith:(NSString *)timeStr{
    if (!timeStr) {
        timeStr = [[NSDate date]LocalTimeISO8601String];
    }
    
    NSCharacterSet *set = [NSCharacterSet characterSetWithCharactersInString:@"-: \n"];
    NSArray *timeArray = [timeStr componentsSeparatedByCharactersInSet:set];
    /* 获得的时间需要精确到秒 元素个数必须大于或者等于6 */
    assert(timeArray.count>=6);
    
    char timeByte[6];
    memset(timeByte, 0, 6);
    NSInteger year = [[timeArray firstObject]integerValue];
    year>1999?(year -= 2000):0;
    timeByte[0] = year;
    for (NSInteger i=1; i<6; i++) {
        
        timeByte[i] = [timeArray[i]integerValue];
    }
    return [NSData dataWithBytes:timeByte length:6];
}
@end

File: BPGBiolands/BPGProtocol.swift
//
//  BPGProtocol.swift
//  BPG-Bioland
//
//  Created by starlueng on 2016/8/12.
//  Copyright © 2016年 starlueng. All rights reserved.
//

import Foundation

enum commonds :UInt16{
    
    case getLastTestData
    case getAllTestData
    case realStateData
    case realTestData
    case realTestMode
    case pauseMode
    case turnOnMode
    case turnOffMode
    case  setTime
    case  getVersion
    case  isSendDeploy
    
    func simpleDescription() -> UInt16 {
        
        switch self {
            
        case .getLastTestData://上传最近一次测量结果命令
            return 0x1131
        case .getAllTestData://上传所有测量结果命令
            return 0x1331
        case .realStateData://时时上传状态数据
            return 0x0831
        case .realTestData://实时测量结果获取命令
            return 0x0931
        case .realTestMode://开始测量
            return 0x2131
        case .pauseMode://停止测量
            return 0x2231
        case .turnOnMode://开机
            return 0x0111
        case .turnOffMode://关机
            return 0x0211
        case .setTime://设置时间
            return 0x0331
        case .isSendDeploy://脉搏数据配置 1.发送 0.不发送
            return 0x0631
        default:
            return 0
        }
    }
    
};

let channelOnCharacteristicView : String = "babyDefault"
let serverID : String                    = "0000FBB0-494C-4F47-4943-544543480000"//服务器id
let readCharacterID : String             = "0000FBB1-494C-4F47-4943-544543480000"//读特征值
let writeCharacterID : String            = "0000FBB2-494C-4F47-4943-544543480000"//写特征值

struct thisCharacteristics {
    
   var write : CBCharacteristic , read :CBCharacteristic
}
class BPGProtocol: AnyObject {

    var baby : BabyBluetooth?,currPeripheral :CBPeripheral?,writerCharacteristic : CBCharacteristic? ,readCharacteristic : CBCharacteristic?,information : Data?
   
     init(){
        

    }
    
    func BLEDelegate()  {
        baby = BabyBluetooth.share()
        self.scanBleWithPrexName(prexName: "serialcom")
    }
    func scanBleWithPrexName(prexName :String) {
        
        //MARK:停止之前的连接
        baby!.cancelAllPeripheralsConnection()
        //MARK:设置委托后直接可以使用，无需等待CBCentralManagerStatePoweredOn状态。
        baby!.scanForPeripherals()()?.begin()();
        
        baby!.setBlockOnCentralManagerDidUpdateState { (CBCentralManager) in
            
            if CBCentralManager?.state == .poweredOn{
            
                print("蓝牙已打开")
            }
        }
        //MARK:1.设置查找设备的过滤器
        baby!.setFilterOnDiscoverPeripherals { (peripheralName, advertisementData, RSSI) -> Bool in
            
            if (peripheralName?.lowercased().hasPrefix(prexName))!{
                
                return true
            }
            return false
        }
        //MARK:2.设置扫描到设备的委托
        baby!.setBlockOnDiscoverToPeripherals() { (central, peripheral, advertisementData, RSSI) in
            
            let peripheralName :String = peripheral!.name!
            if peripheralName.lowercased().hasPrefix(prexName){
            
                self.baby!.cancelScan()
                self.currPeripheral = peripheral
                let hud = MBProgressHUD.showAdded(to: UIApplication.shared.delegate!.window!, animated: true)
                
                self.connectBleWith(thisC: { (this :thisCharacteristics) in
                    
                    hud?.hide(true)
                })

            }
        }
    }
    
    func connectBleWith(thisC: @escaping (thisCharacteristics) ->())  {
        
        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.5) { 
            self.loadData()
        }
        
        let rhythm = BabyRhythm.init()
        //MARK:设置发现设service的Characteristics的委托
        self.baby!.setBlockOnDiscoverCharacteristicsAtChannel(channelOnCharacteristicView, block: { (peripheral, service, error) in
            
            print("===service name:\(service?.uuid)")
            rhythm.beats()
        })
        self.baby!.setBlockOnDisconnect({ (central, peripheral, error) in
            
            print("didDisconnectPeripheral-----\(error)")
            self.loadData()
        })
        //MARK:设置读取characteristics的委托
        self.baby!.setBlockOnReadValueForCharacteristicAtChannel(channelOnCharacteristicView, block: { (peripheral, characteristics, error) in
            
            if characteristics?.uuid.uuidString == "2A23"{
                
                self.information = characteristics?.value!
            }
            if characteristics?.uuid.uuidString == writeCharacterID{
                
                self.writerCharacteristic = characteristics
            }else if  characteristics?.uuid.uuidString == readCharacterID{
                
                self.readCharacteristic = characteristics
            }
            if (self.readCharacteristic != nil) && (self.writerCharacteristic != nil){
                
                let character = thisCharacteristics(write: self.readCharacteristic! , read: self.writerCharacteristic!)
                
                
                thisC(character)
            }
        })
        //MARK:设置写数据成功的block
        self.baby!.setBlockOnDidWriteValueForCharacteristicAtChannel(channelOnCharacteristicView, block: { (characteristic, error) in
            
            print("setBlockOnDidWriteValueForCharacteristicAtChannel characteristic:%@ and new value:\(characteristic?.uuid, characteristic?.value)")
        })

    }
    @objc func loadData() -> Void {
        //MARK:4.开始连接设备
        
            self.baby!.having()(self.currPeripheral)?.and().channel()(channelOnCharacteristicView)?.then().connectToPeripherals()()?.discoverServices()()?.discoverCharacteristics()()?.readValueForCharacteristic()()?.discoverDescriptorsForCharacteristic()()?.readValueForDescriptors()()?.begin()();
    }
    //MARK:断开蓝牙连接
    func babyStopConnect()  {
        
        self.baby!.cancelScan()
        self.baby!.cancelAllPeripheralsConnection()
    }
    //MARK:解析获取的数据存储为model
    func transFormData(data :Data) -> BPGModel? {
        
        let leng = data.count
        var packetData :[UInt8] = [UInt8](repeating:0 ,count :leng)
        let dataStr = String.init(data: data , encoding: String.Encoding.utf8)
        var i = 0
        for ch in dataStr!.utf8 {
            
            packetData[i] = ch
            i += 1
        }
        if (packetData[0] != 0x68) {
            
            return nil;
        }else{
           let model = BPGModel()
            model.head              = packetData[0];
            model.cmd               = packetData[1];
            model.status            = packetData[2];
            let bodyLength          = (UInt16)((packetData[4] << 8) + packetData[5]);
            model.leng              = bodyLength;
            var bodys :[UInt8] = [UInt8](repeating:0 ,count : bodyLength.hashValue)
            for u in 6..<leng-2 {
                
                bodys[u] = packetData[u+6]
            }
            model.body              = bodys
            model.checkcs           = packetData[data.count-2];
            model.end               = packetData[data.count-1];
            
            return model;
        }

    }
    //MARK:发指令收取指令,返回
    func writeValueWith(data : NSData ,success: @escaping (BPGModel) -> (),fail:(NSError?) ->()) {

        if self.currPeripheral?.state == CBPeripheralState.connected {
            
            self.currPeripheral?.writeValue(data as Data, for: self.writerCharacteristic!, type: .withResponse)
            
            if self.currPeripheral == nil {
                
                let error = NSError.init(domain: "蓝牙设备未连接", code: 0, userInfo: nil)
                fail(error)
                return
            }
            if self.currPeripheral?.state != CBPeripheralState.connected {
                
                let error = NSError.init(domain: "蓝牙已经断开连接，请重新连接", code: 0, userInfo: nil)
                fail(error)
                return
            }
        }
        if self.readCharacteristic!.properties == CBCharacteristicProperties.notify || self.readCharacteristic!.properties == CBCharacteristicProperties.indicate {
            
            if ((self.readCharacteristic?.isNotifying) == nil) {
                
                self.currPeripheral?.setNotifyValue(true, for: self.readCharacteristic!)
                self.baby!.notify(self.currPeripheral!, characteristic: self.readCharacteristic!, block: { (peripheral, characteristics, error) in
                    let model = self.transFormData(data: characteristics!.value!)
                    success(model!)
                })
            }else{
            
                let error = NSError.init(domain: "这个characteristic没有nofity的权限", code: 0, userInfo: nil)
                fail(error)
                return
            }
            
        }
    
    }
    //MARK:编辑命令格式
    func writeCommodDataWithcmd(cmd:__uint8_t,common :commonds,body :[UInt8]) ->NSData {
        
        let dataLength = body.count + 9
        let bodyLength = body.count + 2
        var byte :[UInt8] = [UInt8].init(repeating: 0, count: dataLength)
        //(count: dataLength, repeatedValue: 0)
        byte[0] = 0x68;
        byte[1] = cmd;
        byte[3] = UInt8(bodyLength % 256) ;
        byte[4] = UInt8(bodyLength / 256);
        byte[5] = UInt8 (common.simpleDescription() >> 8);
        byte[6] = UInt8 (common.simpleDescription()&0x00ff);
        
        if body.count > 0 {
            
            let bodys = [UInt8].init(repeating: 0, count: body.count)
            for inc in 0 ..< body.count{
                
                
                byte[7+inc] = bodys[inc]
            }
        }
        var calibrate = 0
        for i in 0 ..< (dataLength - 2) {
            
            calibrate += byte[i].hashValue
        }
        byte[dataLength-2] = UInt8(calibrate&0xff)
        byte[dataLength-1] = 0x16
        
        return NSData.init(bytes: byte, length: dataLength)

    }
    //MARK:----------------------命令集=-------------------------------------
    func turnOnModeWith(success:(BPGModel) ->(),fail:(NSError) ->())  {//开机

        let bodyByte = [UInt8].init(repeating: 0, count: 0)
        let valueData = self.writeCommodDataWithcmd(cmd: 0x02, common: .turnOnMode, body: bodyByte)
        self.writeValueWith(data: valueData, success: { (model:BPGModel?) in
            
            print("\(model)")
        }) { (error:NSError?) in
            
                print("\(error)")
        }

    }
    
    func turnOffModeWith(success:(BPGModel) ->(),fail:(NSError) ->()) {//关机
        let bodyByte = [UInt8].init(repeating: 0, count: 0)
        let valueData = self.writeCommodDataWithcmd(cmd: 0x02, common: .turnOffMode, body: bodyByte)
        self.writeValueWith(data: valueData, success: { (model:BPGModel?) in
            
            print("\(model)")
        }) { (error:NSError?) in
            
            print("\(error)")
        }

    }
    func realTestMode(success:(BPGModel) ->(),fail:(NSError) ->()) {//开测
        
        let bodyByte = [UInt8].init(repeating: 0, count: 0)
        let valueData = self.writeCommodDataWithcmd(cmd: 0x02, common: .realTestMode, body: bodyByte)
        self.writeValueWith(data: valueData, success: { (model:BPGModel?) in
            
            print("\(model)")
        }) { (error:NSError?) in
            
            print("\(error)")
        }
        
    }
    func pauseMode(success:(BPGModel) ->(),fail:(NSError) ->()) {//暂停
        
        let bodyByte =  [UInt8].init(repeating: 0, count: 0)
        let valueData = self.writeCommodDataWithcmd(cmd: 0x02, common: .pauseMode, body: bodyByte)
        self.writeValueWith(data: valueData, success: { (model:BPGModel?) in
            
            print("\(model)")
        }) { (error:NSError?) in
            
            print("\(error)")
        }
    }

}
